jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/Content.java

```
@@ -1106,7 +1107,32 @@ static Chunk next(Chunk chunk)
             if (chunk == null)
                 return null;
             if (Content.Chunk.isFailure(chunk))
-                return chunk.isLast() ? chunk : null;
+            {
```

I don't think this is the right place to handle this issue, I don't see anything wrong with `Content.Source.read()` always returning an error chunk that contains the exact same exception instance.

I think this should be handled at the place where the exception is thrown, like in `ContentSinkOutputStream`. Most (all?) places should already use `IO.rethrow()` so we can probably come up with some similar static helper, maybe like `Content.Chunk.throwIfFailure(Chunk)`?

We should also add some tests similar to `ContentSinkOutputStreamTest` for each codepath that throws a chunk's exception.

---

@lorban it is now clear that throwing the same instance is not well handled by the JDK and other libraries.

We cannot expect any user-code wrapper (such as `commons-fileupload`) to handle the case where the exception may be self suppressed or there is a loop.

I don't see any other choice than us throwing a different exception every time, or returning a new failure chunk every time.

I'd keep the code we added in #11759 and #13514, but I'm afraid we have to change the read side.

---

> I don't see any other choice than us throwing a different exception every time, or returning a new failure chunk every time.

Yes, and my point is that we should throw a different exception every time, and NOT return a new failure chunk every time.

---

@lorban not sure what you mean here. The chunk is immutable, so we have to return a new chunk instance, just to change the exception instance inside.

User code (libraries and such) will just do `throw chunk.getFailure()` so we can't really ask them to wrap the failure into a new Throwable, can we?

---

How about `throw chunk.getWrappedFailure(new IOException())` which just makes the existing `chunk.getFailure()` a cause of the one passed in?

---

@sbordet I think we can ask users to not `throw chunk.getFailure()`, especially if we provide a helper.

If we don't and change the failure instance each time `getFailure()` is called, I expect difficulties tracing back a problem where an error chunk has been returned as the stack trace will be lost.

---

But we have the Chunk.next() mechanism entirely to handle what the next chunk to be returned is.   If we read multiple times after a failure then we have to return different chunks if we want different exceptions.    So I think this is exactly the correct place to handle it.

---

I think we need to review `ContentSourceTransformer.read()`, which does not appear to use `next()` properly.

---

@gregw if we do not lose the original cause (following the suggestion @sbordet made in `ExceptionUtil.copyOf()`) then I'm fine with putting the responsibility of making a copy in `Chunk.next()`.

We also have to revisit all our `Content.Source` implementations to make sure they properly call `Chunk.next()`.

---

> @gregw if we do not lose the original cause (following the suggestion @sbordet made in `ExceptionUtil.copyOf()`) then I'm fine with putting the responsibility of making a copy in `Chunk.next()`.

I'm concerned that will still create problems.  But I'll try....
> 
> We also have to revisit all our `Content.Source` implementations to make sure they properly call `Chunk.next()`.

We should do that anyway!